contracts/uniswap_v2/lib.rs

```diff
@@ -0,0 +1,839 @@
+                        .unwrap_or_else(Ratio::max_value);
+                    if price_impact > limit {
+                        return Err(Error::ExceedPriceImpactLimit);
+                    }
+                }
+
+                supply_amounts[i - 1] = supply_amount;
+                i -= 1;
+            }
+
+            Ok(supply_amounts)
+        }
+
+        fn _swap(
+            &mut self,
+            supply_currency_id: TokenAddress,
+            target_currency_id: TokenAddress,
+            supply_increment: Balance,
+            target_decrement: Balance,
+        ) -> Result<()> {
+            if let Some(trading_pair) =
+                TradingPair::from_currency_ids(supply_currency_id, target_currency_id)
+            {
+                let ((pool_0, pool_1), _) = self.pair_info(&trading_pair);
+
+                let invariant_before_swap: U256 =
+                    U256::from(pool_0).saturating_mul(U256::from(pool_1));
+
+                if supply_currency_id == trading_pair.first() {
+                    let pool_0 = pool_0
+                        .checked_add(supply_increment)
+                        .ok_or(Error::ArithmeticOverflow)?;
+                    let pool_1 = pool_1
+                        .checked_sub(target_decrement)
+                        .ok_or(Error::ArithmeticUnderflow)?;
+                    self.liquidity_pool
+                        .insert(trading_pair.clone(), (pool_0, pool_1));
+                } else {
+                    let pool_0 = pool_0
+                        .checked_sub(target_decrement)
+                        .ok_or(Error::ArithmeticUnderflow)?;
+                    let pool_1 = pool_1
+                        .checked_add(supply_increment)
+                        .ok_or(Error::ArithmeticOverflow)?;
+                    self.liquidity_pool
+                        .insert(trading_pair.clone(), (pool_0, pool_1));
+                }
+
+                // invariant check to ensure the constant product formulas (k = x * y)
+                let invariant_after_swap: U256 =
+                    U256::from(pool_0).saturating_mul(U256::from(pool_1));
+                if invariant_after_swap < invariant_before_swap {
+                    return Err(Error::InvariantCheckFailed);
+                }
+                return Ok(());
+            }
+            Ok(())
+        }
+
+        fn _swap_by_path(&mut self, path: &[TokenAddress], amounts: &[Balance]) -> Result<()> {
+            let mut i: usize = 0;
+            while i + 1 < path.len() {
+                let (supply_currency_id, target_currency_id) = (path[i], path[i + 1]);
+                let (supply_increment, target_decrement) = (amounts[i], amounts[i + 1]);
+                self._swap(
+                    supply_currency_id,
+                    target_currency_id,
+                    supply_increment,
+                    target_decrement,
+                )?;
+                i += 1;
```

Comment:
rewrite the loop without the mutable index `i`

!ensure path.len() == amounts.len()